src/llnode_api.cc

```diff
@@ -0,0 +1,227 @@
+      debugger(new lldb::SBDebugger()),
+      target(new lldb::SBTarget()),
+      process(new lldb::SBProcess()),
+      llscan(new LLScan()),
+      v8_(new llnode::v8::LLV8()) {}
+LLNodeApi::~LLNodeApi() = default;
+LLNodeApi::LLNodeApi(LLNodeApi&&) = default;
+LLNodeApi& LLNodeApi::operator=(LLNodeApi&&) = default;
+
+bool LLNodeApi::debugger_initialized_ = false;
+
+/* Initialize the SB API and load the core dump */
+bool LLNodeApi::Init(const char* filename, const char* executable) {
+  if (!LLNodeApi::debugger_initialized_) {
+    lldb::SBDebugger::Initialize();
+    LLNodeApi::debugger_initialized_ = true;
+  }
+
+  if (loaded) {
+    return false;
+  }
+
+  *debugger = lldb::SBDebugger::Create();
+  loaded = true;
+
+  // Single instance target for now
+  *target = debugger->CreateTarget(executable);
+  if (!target->IsValid()) {
+    return false;
+  }
+
+  *process = target->LoadCore(filename);
+  // Load V8 constants from postmortem data
+  v8_->Load(*target);
+  return true;
+}
+
+std::string LLNodeApi::GetProcessInfo() {
+  lldb::SBStream info;
+  process->GetDescription(info);
+  return std::string(info.GetData());
+}
+
+uint32_t LLNodeApi::GetProcessID() { return process->GetProcessID(); }
+
+std::string LLNodeApi::GetProcessState() {
+  int state = process->GetState();
+
+  switch (state) {
+    case lldb::StateType::eStateInvalid:
+      return "invalid";
+    case lldb::StateType::eStateUnloaded:
+      return "unloaded";
+    case lldb::StateType::eStateConnected:
+      return "connected";
+    case lldb::StateType::eStateAttaching:
+      return "attaching";
+    case lldb::StateType::eStateLaunching:
+      return "launching";
+    case lldb::StateType::eStateStopped:
+      return "stopped";
+    case lldb::StateType::eStateRunning:
+      return "running";
+    case lldb::StateType::eStateStepping:
+      return "stepping";
+    case lldb::StateType::eStateCrashed:
+      return "crashed";
+    case lldb::StateType::eStateDetached:
+      return "detached";
+    case lldb::StateType::eStateExited:
+      return "exited";
+    case lldb::StateType::eStateSuspended:
+      return "suspended";
+    default:
+      return "unknown";
+  }
+}
+
+uint32_t LLNodeApi::GetThreadCount() { return process->GetNumThreads(); }
+
+uint32_t LLNodeApi::GetFrameCount(size_t thread_index) {
+  lldb::SBThread thread = process->GetThreadAtIndex(thread_index);
+  if (!thread.IsValid()) {
+    return 0;
+  }
+  return thread.GetNumFrames();
+}
+
+// TODO: should return a class with
+// functionName, directory, file, complieUnitDirectory, compileUnitFile
+std::string LLNodeApi::GetFrame(size_t thread_index, size_t frame_index) {
+  lldb::SBThread thread = process->GetThreadAtIndex(thread_index);
+  lldb::SBFrame frame = thread.GetFrameAtIndex(frame_index);
+  lldb::SBSymbol symbol = frame.GetSymbol();
+
+  std::string result;
+  char buf[4096];
+  if (symbol.IsValid()) {
+    snprintf(buf, sizeof(buf), "Native: %s", frame.GetFunctionName());
```

Comment:
I kept the `snprintf` because with the string streams the output format is not as clear.